src/llmq/quorums.cpp

```diff
@@ -502,14 +500,50 @@ std::vector<CQuorumCPtr> CQuorumManager::ScanQuorums(Consensus::LLMQType llmqTyp
         return {};
     }
 
-    const CBlockIndex* pIndexScanCommitments{pindexStart};
+    const CBlockIndex* pindexStore{pindexStart};
+    const auto& llmq_params_opt = GetLLMQParams(llmqType);
+    assert(llmq_params_opt.has_value());
+
+    {
+        // Quorum sets can only change during the mining phase of DKG.
+        // Find the closest known block index.
+        const int quorumCycleStartHeight = pindexStart->nHeight - (pindexStart->nHeight % llmq_params_opt->dkgInterval);
+        const int quorumCycleMiningStartHeight = quorumCycleStartHeight + llmq_params_opt->dkgMiningWindowStart;
+        const int quorumCycleMiningEndHeight = quorumCycleStartHeight + llmq_params_opt->dkgMiningWindowEnd;
+
+        if (pindexStart->nHeight < quorumCycleMiningStartHeight) {
+            // too early for this cycle, use the previous one
+            pindexStore = pindexStart->GetAncestor(quorumCycleMiningEndHeight - llmq_params_opt->dkgInterval);
+        } else if (pindexStart->nHeight > quorumCycleMiningEndHeight) {
+            // we are past the mining phase of this cycle, use it
+            pindexStore = pindexStart->GetAncestor(quorumCycleMiningEndHeight);
+        }
+        // everything else is inside the mining phase of this cycle, no pindexStore adjustment needed
+
+        if (pindexStore == nullptr || !utils::IsQuorumTypeEnabled(llmqType, *this, pindexStore)) {
+            return {};
+        }
+    }
+
+    const CBlockIndex* pIndexScanCommitments{pindexStore};
     size_t nScanCommitments{nCountRequested};
     std::vector<CQuorumCPtr> vecResultQuorums;
 
     {
         LOCK(cs_scan_quorums);
+        if (scanQuorumsCache.empty()) {
+            for (const auto& llmq : Params().GetConsensus().llmqs) {
+                // NOTE: We store it for each block hash in the DKG mining phase here
+                // and not for a single quorum hash per quorum like we do for other caches.
+                // And we only do this for max_cycles() of the most recent quorums
+                // because signing by old quorums requires the exact quorum hash to be specified
+                // and quorum scanning isn't needed there.
+                scanQuorumsCache.emplace(std::piecewise_construct, std::forward_as_tuple(llmq.type),
```

Comment:
also; maybe try_emplace

Reply:
clang-tidy / sonarlint will be annoyed about `emplace` anyhow and will want it changed to try_emplace

Reply:
I wonder what's the exact warning you get from these tools and why `try_emplace` would be better here?

Reply:
see https://gist.github.com/PastaPastaPasta/b16313c79bfcb9cd5ae7667c00623011

Reply:
Interesting. Applied.

Reply:
Yeah; I don't know how valid it is; but the rule probably exists for a reason 🤷